### ISSUE TITLE
Converted datatype not using type from dialect after calling nullable

### ISSUE THREAD

**Issue by juriad** [jOOQ/jOOQ#10312]
### Expected behavior and actual behavior:

Postgres does not support `tinyint`; it uses `smallint` instead. However, some columns of the following example are generated as `tinyint`. The example also shows a possible workaround (set `nullable` before converting the type).

I think a bug is somewhere in `org.jooq.impl.DefaultDataType#getDataType(org.jooq.Configuration)`. After calling `nullable(...)`, the datatype stops being `ConveredDataType` and switches to `DefaultDataType`.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

Sorry, the code is just so short that I copy/paste it here.

```
public class X {
    enum E {
        F
    }

    public static void main(String[] args) {
        String sql = DSL.using(SQLDialect.POSTGRES)
                .createTable("t")
                .column("cnn", SQLDataType.TINYINT.nullable(false))
                .column("cn", SQLDataType.TINYINT.nullable(true))
                .column("c", SQLDataType.TINYINT)
                .column("enn", SQLDataType.TINYINT.asConvertedDataType(new EnumConverter<>(Byte.class, E.class)).nullable(false))
                .column("en", SQLDataType.TINYINT.asConvertedDataType(new EnumConverter<>(Byte.class, E.class)).nullable(true))
                .column("e", SQLDataType.TINYINT.asConvertedDataType(new EnumConverter<>(Byte.class, E.class)))
                .column("nne", SQLDataType.TINYINT.nullable(false).asConvertedDataType(new EnumConverter<>(Byte.class, E.class)))
                .column("ne", SQLDataType.TINYINT.nullable(true).asConvertedDataType(new EnumConverter<>(Byte.class, E.class)))
                .getSQL();
        System.out.println(sql);
    }
}
```

Prints:
```
create table "t"("cnn" smallint not null, "cn" smallint null, "c" smallint null, "enn" tinyint not null, "en" tinyint null, "e" smallint null, "nne" smallint not null, "ne" smallint null)
```

### Versions:

- jOOQ: jooq-3.13.1.jar (pro)
- Java: 11.0.6
- Database (include vendor): none
- OS: Linux
- JDBC Driver (include name if inofficial driver): none

**Comment by lukaseder** [comment]
Thanks for the thorough report, nice catch!

**Comment by lukaseder** [comment]
The fix could profit from the new `AbstractDataType` which was introduced with #9061, and allows for a cleaner `ConvertedDataType` implementation

**Comment by lukaseder** [comment]
Letting `ConvertedDataType` extend `AbstractDataType` instead of `DefaultDataType` prevents custom data types from being added to the internal data type cache, which means that doing things like `inline(E.F)` no longer works:

```
Caused by: org.jooq.exception.SQLDialectNotSupportedException: Type class E is not supported in dialect DEFAULT
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:628)
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:570)
	at org.jooq.impl.DSL.getDataType(DSL.java:26120)
	at org.jooq.impl.DSL.val(DSL.java:23816)
	at org.jooq.impl.DSL.inline(DSL.java:22892)
```

While it is not desireable to have custom data types in this static type cache, this will still cause a few regressions in client code, if not properly handled. I'll have to think about this.

**Comment by lukaseder** [comment]
So, this is a prerequisite: https://github.com/jOOQ/jOOQ/issues/10438

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14. The fix requires a few changes to how converted data types are managed, which risks regressions when people rely on the internal static type registry (see e.g. https://github.com/jOOQ/jOOQ/issues/10438#issuecomment-662923207). Given that the workarounds are simple, i.e. just reorder the method calls, I will not backport this fix to 3.13.
